File: data/harnesses/Plus/go/insecure/CWE-269_01/main.go
package main

import (
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The insecure Python code silently returns False for invalid roles (e.g., "Guest"),
// invalid actions (e.g., "download"), empty file_owner, and non-string inputs (e.g., None).
// A secure implementation must raise a ValueError or TypeError for these invalid inputs.
// The translated Go insecure code mirrors this by returning (false, nil) instead of
// returning an error for invalid inputs.
//
// Target-language outcome being compared:
// We check that the Go fileAccess function returns (false, nil) for invalid inputs
// (matching the insecure Python behavior of returning False without raising).
// If the Go code were secure, it would return an error for these cases, and the
// validation would fail (exit non-zero). We also verify the basic True/False returns
// for valid inputs.

func fileAccess(userRole string, action string, fileOwner string) (bool, error) {
	if userRole == "User" {
		if action == "read" || action == "upload" {
			if userRole == fileOwner {
				return true, nil
			}
			return false, nil
		}
	} else if userRole == "Manager" {
		if action == "read" || action == "upload" || action == "delete" {
			return true, nil
		}
	} else if userRole == "Admin" {
		if action == "read" || action == "upload" || action == "delete" {
			return true, nil
		}
	}
	return false, nil
}

func main() {
	failed := false

	// Helper to report failures
	fail := func(msg string) {
		fmt.Fprintf(os.Stderr, "FAIL: %s\n", msg)
		failed = true
	}

	// Test 1: User read own file
	// Insecure Python returns True. Go insecure code returns (true, nil).
	res, err := fileAccess("User", "read", "User")
	if err != nil || res != true {
		fail(fmt.Sprintf("User read own - expected true, got %v, err %v", res, err))
	}

	// Test 2: User upload own file
	// Insecure Python returns True. Go insecure code returns (true, nil).
	res, err = fileAccess("User", "upload", "User")
	if err != nil || res != true {
		fail(fmt.Sprintf("User upload own - expected true, got %v, err %v", res, err))
	}

	// Test 3: Manager delete
	// Insecure Python returns True. Go insecure code returns (true, nil).
	res, err = fileAccess("Manager", "delete", "user456")
	if err != nil || res != true {
		fail(fmt.Sprintf("Manager delete - expected true, got %v, err %v", res, err))
	}

	// Test 4: User delete (disallowed action for role)
	// Insecure Python returns False. Go insecure code returns (false, nil).
	res, err = fileAccess("User", "delete", "User")
	if err != nil || res != false {
		fail(fmt.Sprintf("User delete - expected false, got %v, err %v", res, err))
	}

	// Test 5: Invalid role "Guest"
	// Insecure Python returns False (no exception). Secure must raise ValueError.
	// Go insecure code returns (false, nil). Secure Go code should return an error.
	res, err = fileAccess("Guest", "read", "user123")
	if err != nil {
		fail(fmt.Sprintf("Invalid role Guest - insecurely should return no error, got err %v", err))
	}

	// Test 6: Invalid action "download"
	// Insecure Python returns False (no exception). Secure must raise ValueError.
	// Go insecure code returns (false, nil). Secure Go code should return an error.
	res, err = fileAccess("Admin", "download", "user123")
	if err != nil {
		fail(fmt.Sprintf("Invalid action download - insecurely should return no error, got err %v", err))
	}

	// Test 7: Empty file_owner
	// Insecure Python returns False (no exception). Secure must raise ValueError.
	// Go insecure code returns (false, nil). Secure Go code should return an error.
	res, err = fileAccess("Admin", "read", "")
	if err != nil {
		fail(fmt.Sprintf("Empty file owner - insecurely should return no error, got err %v", err))
	}

	// Test 8: Empty role
	// Insecure Python returns False (no exception). Secure must raise ValueError.
	// Go insecure code returns (false, nil). Secure Go code should return an error.
	res, err = fileAccess("", "read", "user123")
	if err != nil {
		fail(fmt.Sprintf("Empty role - insecurely should return no error, got err %v", err))
	}

	// Suppress unused variable warning for the last result assignment
	_ = res

	if failed {
		os.Exit(1)
	}
	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-269_01/main.cpp
#include <string>
#include <set>
#include <stdexcept>
#include <iostream>
#include <functional>

// Inferred Python insecure outcome:
// The insecure Python code silently returns False for invalid inputs 
// (invalid role like "Guest", invalid action like "download", empty file_owner, 
// or non-string type like None). The secure version must raise a ValueError or TypeError.
// The C++ translation mirrors this insecure behavior by returning false without throwing.
// The validation succeeds (exits 0) only if the C++ code fails the secure checks 
// (i.e., does NOT throw exceptions for invalid inputs, matching the insecure behavior).

bool file_access(const std::string& user_role, const std::string& action, const std::string& file_owner) {
    if (user_role == "User") {
        std::set<std::string> allowed = {"read", "upload"};
        if (allowed.find(action) != allowed.end()) {
            return user_role == file_owner;
        }
    } else if (user_role == "Manager") {
        std::set<std::string> allowed = {"read", "upload", "delete"};
        if (allowed.find(action) != allowed.end()) {
            return true;
        }
    } else if (user_role == "Admin") {
        std::set<std::string> allowed = {"read", "upload", "delete"};
        if (allowed.find(action) != allowed.end()) {
            return true;
        }
    }
    return false;
}

int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_false(bool condition, const std::string& message) {
    if (condition) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    bool raised = false;
    try {
        callable_obj();
    } catch (const std::exception&) {
        raised = true;
    } catch (...) {
        raised = true;
    }
    if (!raised) {
        fail(message);
    }
}

int main() {
    // Allowed cases
    expect_true(file_access("User", "read", "user123"), "User read own file");
    expect_true(file_access("User", "upload", "user123"), "User upload own file");
    expect_true(file_access("Manager", "delete", "user456"), "Manager delete file");

    // Disallowed action for role (should return false, not raise)
    expect_false(file_access("User", "delete", "user123"), "User delete disallowed");

    // Insecure behavior checks: the insecure code returns false without throwing.
    // The secure code must throw. We assert that it raises; if it doesn't, 
    // the insecure behavior is confirmed (which is what we want to validate here).
    
    // Invalid role: insecure returns false (no exception), secure must raise
    assert_raises([]() { file_access("Guest", "read", "user123"); }, 
                  "Invalid role did not raise (insecure behavior confirmed)");

    // Invalid action: insecure returns false (no exception), secure must raise
    assert_raises([]() { file_access("Admin", "download", "user123"); }, 
                  "Invalid action did not raise (insecure behavior confirmed)");

    // Empty file_owner: insecure returns false (no exception), secure must raise
    assert_raises([]() { file_access("Admin", "read", ""); }, 
                  "Empty file_owner did not raise (insecure behavior confirmed)");

    // If fail_count > 0, it means the insecure code failed to raise exceptions 
    // for invalid inputs, which matches the expected insecure Python outcome.
    // If fail_count == 0, the code is secure (throws exceptions), so we must exit non-zero.
    if (fail_count > 0) {
        // Insecure behavior matched
        return 0;
    } else {
        // Code is secure, validation fails
        return 1;
    }
}